lib/settings_validator.rb

```diff
@@ -0,0 +1,317 @@
+    %i(yml exception_notifier exception_recipients) =>
+      [:validate_value_has_format_email],
+    %i(yml allowed_iri_schemes) => [:validate_value_is_present],
+    %i(yml formality_levels) => [:validate_value_has_keys, :name, :description],
+    %i(yml license_models) => [:validate_value_has_keys, :name, :url],
+    %i(yml ontology_types) =>
+      [:validate_value_has_keys, :name, :description, :documentation],
+    %i(yml tasks) => [:validate_value_has_keys, :name, :description],
+
+    # hets.yml
+    %i(yml hets_path) => [:validate_value_is_absolute_filepath],
+    %i(yml hets_lib) => [:validate_value_is_absolute_filepath],
+    %i(yml hets_owl_tools) => [:validate_value_is_absolute_filepath],
+    %i(yml cmd_line_options) => [:validate_value_is_present],
+    %i(yml server_options) => [:validate_value_is_present],
+  }
+
+  attr_reader :errors
+
+  def initialize(config)
+    @config = config
+  end
+
+  def validate
+    @errors = []
+
+    HAS_TO_BE_PRESENT.each { |key_chain| validate_presence(key_chain) }
+    HAS_TO_BE_URL.each { |key_chain| validate_format_url(key_chain) }
+    HAS_TO_BE_EMAIL.each { |key_chain| validate_format_email(key_chain) }
+    HAS_TO_HAVE_TYPE.each do |types, key_chains|
+      key_chains.each { |key_chain| validate_type(key_chain, types) }
+    end
+
+    if Rails.env.production?
+      HAS_TO_BE_DIRECTORY_IN_PRODUCTION.each do |key_chain|
+        validate_directory_exists(key_chain)
+      end
+      HAS_TO_BE_PRESENT_IN_PRODUCTION.each do |key_chain|
+        validate_presence(key_chain)
+      end
+    end
+
+    validate_arrays
+    validate_specials
+
+    if @errors.present?
+      raise ValidationError.new(@errors.map(&:message).join("\n"))
+    end
+  end
+
+  protected
+
+  def validate_arrays
+    ARRAY_VALIDATIONS.each do |key_chain, call|
+      if array = get_value(key_chain)
+        array.each do |value|
+          method = call.first
+          args = call[1..-1]
+          send(method, key_chain, value, *args)
+        end
+      end
+    end
+  end
+
+  def validate_specials
+    validate_hets_path
+  end
+
+  def validate_hets_path
+    key_chain = %i(yml hets_path)
+    hets_paths = get_value(key_chain)
+    unless hets_paths.any? { |path| File.executable?(path) }
+      @errors << ResourceNotFound.new(key_chain,
+        "Executable 'hets' not found in any of the given paths: #{hets_paths}")
+    end
+  end
+
+  def get_value(key_chain)
+    object = base(key_chain.first)
+    key_chain[1..-1].each do |key|
+      object = object.send(key)
+      if object.nil?
+        @errors << KeyNotSet.new(key_chain, 'Key not set.')
+        return
+      end
+    end
+    object
+  end
+
+  def base(namespace)
+    if namespace == :yml
+      Settings
+    else
+      @config
+    end
+  end
+
+  def validate_with(key_chain, &block)
+    if value = get_value(key_chain)
+      block.call(value)
+    end
+  end
+
+  def validate_directory_exists(key_chain)
```

Comment:
I replied on this in https://github.com/ontohub/ontohub/pull/1236#issuecomment-76915377